optuna/_gp/_search_space.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+
+import math
+from typing import Any
+from typing import NamedTuple
+
+import numpy as np
+import scipy.stats.qmc
+
+import optuna
+
+
+LINEAR = 0
+LOG = 1
+CATEGORICAL = 2
+
+
+class SearchSpace(NamedTuple):
+    param_type: np.ndarray
+    bounds: np.ndarray
+    step: np.ndarray
+
+
+def untransform_one_param(
+    x: np.ndarray, param_type: int, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    if param_type == CATEGORICAL:
+        return x
+    else:
+        bounds2 = (bounds[0] - 0.5 * step, bounds[1] + 0.5 * step)
+        if param_type == LOG:
+            bounds2 = (math.log(bounds2[0]), math.log(bounds2[1]))
+        x = x * (bounds2[1] - bounds2[0]) + bounds2[0]
+        if param_type == LOG:
+            x = np.exp(x)
+        return x
+
+
+def transform_one_param(
+    x: np.ndarray, param_type: int, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    if param_type == CATEGORICAL:
+        return x
+    else:
+        bounds2 = (bounds[0] - 0.5 * step, bounds[1] + 0.5 * step)
+        if param_type == LOG:
+            bounds2 = (math.log(bounds2[0]), math.log(bounds2[1]))
+            x = np.log(x)
+        x = (x - bounds2[0]) / (bounds2[1] - bounds2[0])
+        return x
+
+
+def round_one_transformed_param(
+    x: np.ndarray, param_type: int, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    assert param_type != CATEGORICAL
+    if step == 0.0:
+        return x
+
+    x = untransform_one_param(x, param_type, bounds, step)
+    x = (x - bounds[0] + 0.5 * step) // step * step + bounds[0]
+    x = transform_one_param(x, param_type, bounds, step)
+    return x
+
+
+def sample_transformed_params(n: int, search_space: SearchSpace) -> np.ndarray:
+    dim = search_space.param_type.shape[0]
+    param_types, bounds, steps = search_space
+    qmc_engine = scipy.stats.qmc.Sobol(dim, scramble=True)
+    xs = qmc_engine.random(n)
+    for i in range(dim):
+        if param_types[i] == CATEGORICAL:
+            xs[:, i] = np.floor(xs[:, i] * bounds[i, 1])
+        elif steps[i] != 0.0:
+            xs[:, i] = round_one_transformed_param(
+                xs[:, i], param_types[i], (bounds[i, 0], bounds[i, 1]), steps[i]
+            )
+    return xs
+
+
+def get_search_space_and_transformed_params(
+    trials: list[optuna.trial.FrozenTrial],
+    optuna_search_space: dict[str, optuna.distributions.BaseDistribution],
+) -> tuple[SearchSpace, np.ndarray]:
```

Comment:
Could you check if we can use `_SearchSpaceTransform` [here](https://github.com/optuna/optuna/blob/master/optuna/_transform.py#L16)?

Reply:
No, `_SearchSpaceTransform` in its current form cannot be used here because it converts categorical parameters to one-hot vectors. How about leaving it a follow-up to merge this file with `_SearchSpaceTransform`?